### DISCUSSION TITLE
Split pane mode with shell and nav in sync?

### DISCUSSION THREAD

**Discussion by Urist-McUristurister** [jarun/nnn#1674] [General]
Is it possible to run `nnn` in a kind of a split mode, where the shell is in left pane, and nnn is in the right, and they're synchronised? I.e. when changing directory with `cd` (or otherwise) in left pane, this change is reflected in `nnn` in the right pane. Correspondingly, if I navigate through directories in `nnn` in the right pane, and then switch to the left ("shell") pane, the pwd in the left pane is changed to the current directory of `nnn` in the right pane.

Now that I re-read it, it sounds intricate, so I'll try to explain it in pseudo-code...

```ruby
shell_pane.on_event("pwd_change") do
  nnn_pane.cd(shell_pane.pwd)
done

shell_pane.on_event("became_active") do
  if shell_pane.pwd != nnn_pane.pwd then
    shell_pane.cd(nnn_pane.pwd)
  endif
done
```

Hope that's more clear...

Is it possible to achieve such behaviour? I'm using iTerm2 if that matters.

**Comment by jarun** [comment]
Sorry, `nnn` doesn't support split pane natively.

**Reply by Urist-McUristurister** [reply]
I figured, hence the **discussion**.

Maybe it is possible to achieve this via plugins, since, according to plugins README, `nnn provides a mechanism for plugins to send data to nnn to control its active directory`.
Maybe there is a way to perform the same in opposite direction (ask nnn for it's current folder) that I've missed.
Maybe someone of the nnn users did something like that or even remotely similar — you never know.

Even if "nnn doesn't support split pane natively", it doesn't mean the discussion should be immediately closed.

**Reply by jarun** [reply]
Dual-pane is a massive work. In my experience I haven't found too many people who have such time to invest in `nnn`.

**Reply by jarun** [reply]
If anyone from the team is interested they would respond back anyway.

**Reply by N-R-K** [reply]
> Maybe there is a way to perform the same in opposite direction (ask nnn for it's current folder) that I've missed.

There are some X11 related code (which are enabled with flag `-x`) that uses OSC-7 to signal CWD changes to the terminal. Also, when `-x` is enabled, nnn sets the terminal window's title to the CWD.
 
https://github.com/jarun/nnn/blob/26d5b5c6141f0424869db8ca47e1d370f7b898d5/src/nnn.c#L1358-L1366

Maybe you can make use of this as a hacky workaround to get what you want.

**Reply by Urist-McUristurister** [reply]
> Dual-pane is a massive work. In my experience I haven't found too many people who have such time to invest in `nnn`.

I would like to bring your attention to the fact that nowhere in my post I've mentioned a request to implement such a feature.

This is not a feature request.

Not a complaint.

Not a bug.

Not an **issue**.

This is a request for a **discussion**.

And I must admit, I'm fairly surprised by the fact that discussions are considered to be something that is limited to the post's author and the team exclusively.

**Reply by Urist-McUristurister** [reply]
> > Maybe there is a way to perform the same in opposite direction (ask nnn for it's current folder) that I've missed.
> 
> There are some X11 related code (which are enabled with flag `-x`) that uses OSC-7 to signal CWD changes to the terminal. Also, when `-x` is enabled, nnn sets the terminal window's title to the CWD.
> 
> https://github.com/jarun/nnn/blob/26d5b5c6141f0424869db8ca47e1d370f7b898d5/src/nnn.c#L1358-L1366
> 
> Maybe you can make use of this as a hacky workaround to get what you want.

Thanks! That indeed helps, at least I now have a direction to dig. ;)

**Comment by pvxx** [comment]
If you use tmux, you can use `send-keys` as a simple but possibly dangerous solution. Start with tmux split window opening nnn in one pane and shell in another, then use custom plugin in nnn to send `cd {current dir}` into shell-pane and some command to set it in nnn. Not seamless but may be enough for what you need.

I personally only use a simple plugin to open split pane with shell from nnn (so I can keep browsing the filesystem while running commands and see both), never had the need to send directory from shell "back" ;). But I use `send-keys` in specific work case where I get URLs in text-file, that I have to run semi-interactive scripts on, it is more convenient than to copy-paste all the time ;).
